rule target:
    input:
        "data/qiime/total_sum_scaling.tsv",
        "data/qiime/total_sum_scaling.biom",
        "data/qiime/total_sum_scaling.qza",
        "data/misc/dna-sequences.fasta",
        "databases/sepp-refs-silva-128.qza",
        "databases/silva-138-99-515-806-nb-classifier.qza",
        "data/qiime/tree.qza",
        "data/qiime/placements.qza",
        "data/qiime/total_sum_filt_table.qza",
        "data/qiime/total_sum_rem_table.qza",
        "data/qiime/taxonomy.qza",
        "data/qiime/taxonomy_filtered.qza",
        "data/qiime/taxonomy_filtered.qzv",
        "data/qiime/tax_filt_actual.qza",
        "data/qiime/core_outputs",
        "data/picrust/out_pipeline",
        "data/picrust/tss3_meta_contrib.tsv",
        "data/misc/tss_seq_depth.tsv",
        "data/misc/processed_metadata.tsv",
        "plots/unweighted_unifrac_pcoa.pdf",
        "plots/weighted_unifrac_pcoa.pdf",
        "stats/w_adonis_results.tsv",
        "stats/uw_adonis_results.tsv",
        "plots/faith_pd.pdf",
        "plots/shannon_entropy.pdf",
        "stats/faith_total_results.tsv",
        "stats/faith_diet_results.tsv",
        "stats/shannon_total_results.tsv",
        "stats/shannon_diet_results.tsv",
        "stats/wu_homogeneity.tsv",
        "stats/uu_homogeneity.tsv",
        "plots/wu_homogeneity.pdf",
        "plots/uu_homogeneity.pdf",
        "stats/uu_resiliency.tsv",
        "stats/wu_resiliency.tsv",
        "plots/wu_resiliency.pdf",
        "plots/uu_resiliency.pdf",
        "plots/family_abun1.pdf",
        "plots/family_abun2.pdf",
        "stats/family_abun_lm.tsv",
        "plots/histopathology.pdf",
        "stats/histopathology_lm.tsv",
        "stats/histopathology_dunn.tsv",
        "plots/butyrate_kinase.pdf",
        "plots/butyryl_coa_transferase.pdf",
        "plots/baiH.pdf",
        "plots/baiI.pdf",
        "stats/buty_enzyme_lm.tsv",
        "stats/bile_enzyme_lm.tsv"


rule total_sum_scaling:
    input:
        biom_fp = "data/misc/euk_filt_mergedDietAim1table_051523-Copy1.qza",
        lacto_asv_fp = "data/misc/lactoOnlydna-sequences.fasta"
    output:
        "data/qiime/total_sum_scaling.tsv"
    shell:
        """
        Rscript scripts/total_sum_scaling.R
        """


rule tss_tsv2biom:
    input:
       "data/qiime/total_sum_scaling.tsv" 
    output:
        "data/qiime/total_sum_scaling.biom"
    conda:
        "envs/qiime2.yml"
    shell:
        """
        biom convert \
            -i ./data/qiime/total_sum_scaling.tsv \
            -o ./data/qiime/total_sum_scaling.biom \
            --table-type "Table" \
            --to-hdf5
        """


rule tss_biom2qza:
    input:
        "data/qiime/total_sum_scaling.biom"
    output:
        "data/qiime/total_sum_scaling.qza"
    conda:
        "envs/qiime2.yml"
    shell:
        """
        qiime tools import \
            --input-path ./data/qiime/total_sum_scaling.biom \
            --type 'FeatureTable[Frequency]' \
            --output-path ./data/qiime/total_sum_scaling.qza
        """


rule rep_seqs2fasta:
    input:
        "data/misc/euk_filt-mergedDietAim1rep-seqs_051523-Copy1.qza"
    output:
        "data/misc/dna-sequences.fasta"
    conda:
        "envs/qiime2.yml"
    shell:
        """
        qiime tools export \
            --input-path ./data/misc/euk_filt-mergedDietAim1rep-seqs_051523-Copy1.qza \
            --output-path ./data/misc
        """


rule get_reference_databases:
    output:
        "databases/sepp-refs-silva-128.qza",
        "databases/silva-138-99-515-806-nb-classifier.qza"
    shell:
        """
        wget https://data.qiime2.org/2023.5/common/sepp-refs-silva-128.qza -P ./databases/
        wget https://data.qiime2.org/2020.8/common/silva-138-99-515-806-nb-classifier.qza -P./databases/
        """


rule sepp_phylo_tree:
    input:
        "data/misc/euk_filt-mergedDietAim1rep-seqs_051523-Copy1.qza",
        "databases/sepp-refs-silva-128.qza"
    output:
        "data/qiime/tree.qza",
        "data/qiime/placements.qza"
    conda:
        "envs/qiime2.yml"
    shell:
        """
        qiime fragment-insertion sepp \
            --i-representative-sequences ./data/misc/euk_filt-mergedDietAim1rep-seqs_051523-Copy1.qza \
            --i-reference-database ./databases/sepp-refs-silva-128.qza \
            --o-tree ./data/qiime/tree.qza \
            --o-placements ./data/qiime/placements.qza
        """


rule sepp_ASV_filtering:
    input:
        "data/qiime/total_sum_scaling.qza",
        "data/qiime/tree.qza"
    output:
        "data/qiime/total_sum_filt_table.qza",
        "data/qiime/total_sum_rem_table.qza"
    conda:
        "envs/qiime2.yml"
    shell:
        """
        qiime fragment-insertion filter-features \
            --i-table ./data/qiime/total_sum_scaling.qza \
            --i-tree ./data/qiime/tree.qza \
            --o-filtered-table ./data/qiime/total_sum_filt_table.qza \
            --o-removed-table ./data/qiime/total_sum_rem_table.qza
        """


rule taxonomic_classification:
    input:
        "databases/silva-138-99-515-806-nb-classifier.qza",
        "data/misc/euk_filt-mergedDietAim1rep-seqs_051523-Copy1.qza"
    output:
        "data/qiime/taxonomy.qza"
    conda:
        "envs/qiime2.yml"
    shell:
        """
        qiime feature-classifier classify-sklearn \
            --i-classifier ./databases/silva-138-99-515-806-nb-classifier.qza \
            --i-reads ./data/misc/euk_filt-mergedDietAim1rep-seqs_051523-Copy1.qza \
            --o-classification ./data/qiime/taxonomy.qza
        """


rule filter_taxonomy:
    input:
        "data/qiime/total_sum_scaling.qza",
        "data/qiime/taxonomy.qza"
    output:
        "data/qiime/taxonomy_filtered.qza"
    conda:
        "envs/qiime2.yml"
    shell:
        """
        qiime taxa filter-table \
            --i-table ./data/qiime/total_sum_scaling.qza \
            --i-taxonomy ./data/qiime/taxonomy.qza \
            --p-include p_ \
            --p-exclude mitochondria,chloroplast \
            --o-filtered-table ./data/qiime/taxonomy_filtered.qza
        """


rule filter_taxonomy_vis:
    input:
        "data/qiime/taxonomy_filtered.qza"
    output:
        "data/qiime/taxonomy_filtered.qzv"
    conda:
        "envs/qiime2.yml"
    shell:
        """
        qiime feature-table summarize \
            --i-table ./data/qiime/taxonomy_filtered.qza \
            --o-visualization ./data/qiime/taxonomy_filtered.qzv
        """


rule pre_core_metrics_filter:
    input:
        "data/qiime/taxonomy_filtered.qza",
        "data/misc/merged_metadata1.tsv"
    output:
        "data/qiime/tax_filt_actual.qza"
    conda:
        "envs/qiime2.yml"
    shell:
        """
        qiime feature-table filter-samples \
            --i-table ./data/qiime/taxonomy_filtered.qza \
            --m-metadata-file ./data/misc/merged_metadata1.tsv \
            --o-filtered-table ./data/qiime/tax_filt_actual.qza
        """


rule core_metrics_analysis:
##  NEED TO CHECK THE SAMPLING DEPTH BEFORE YOU RUN THIS STEP
    input:
        "data/qiime/tree.qza",
        "data/qiime/taxonomy_filtered.qza",
        "data/misc/merged_metadata1.tsv"
    output:
        "data/qiime/core_outputs"
    conda:
        "envs/qiime2.yml"
    shell:
        """
        qiime diversity core-metrics-phylogenetic \
            --i-phylogeny ./data/qiime/tree.qza \
            --i-table ./data/qiime/taxonomy_filtered.qza \
            --p-sampling-depth  99976 \
            --m-metadata-file ./data/misc/merged_metadata1.tsv \
            --output-dir ./data/qiime/core_outputs
        """


rule picrust2:
    input:
       "data/misc/dna-sequences.fasta",
       "data/qiime/total_sum_scaling.biom"
    output:
       "data/picrust/out_pipeline"
    conda:
        "envs/picrust2.yml"
    shell:
        """
        picrust2_pipeline.py \
            -s ./data/misc/dna-sequences.fasta \
            -i ./data/qiime/total_sum_scaling.biom \
            -o ./data/picrust/out_pipeline \
            --stratified \
            --per_sequence_contrib \
            -p 32
        """


rule ko_contrib_filter:
    input:
        ko_in = "data/picrust/out_pipeline/KO_metagenome_out/pred_metagenome_contrib.tsv.gz"
    output:
        ko_out = "data/picrust/tss3_meta_contrib.tsv"
    shell:
        """
        Rscript scripts/ko_contrib_filter.R
        """


rule sequencing_depth_calculation:
    input:
        biom_fp = "data/misc/euk_filt_mergedDietAim1table_051523-Copy1.qza",
        lacto_asv_fp = "data/misc/lactoOnlydna-sequences.fasta"
    output:
        "data/misc/tss_seq_depth.tsv"
    shell:
        """
        Rscript scripts/seq_depth.R
        """


rule metadata_processing:
    input:
        metadata_FP = "data/misc/merged_metadata1.tsv",
        seq_depth_FP = "data/misc/tss_seq_depth.tsv"
    output:
        "data/misc/processed_metadata.tsv"
    shell:
        """
        Rscript scripts/metadata_processing.R
        """


rule alpha_diversity_plots:
    input:
        metadata_FP = "data/misc/processed_metadata.tsv",
        faith_pd_FP = "data/qiime/core_outputs/faith_pd.tsv",
        shannon_FP = "data/qiime/core_outputs/shannon_entropy.tsv"
    output:
        "plots/faith_pd.pdf",
        "plots/shannon_entropy.pdf"
    shell:
        """
        Rscript scripts/alpha_div_plots.R
        """


rule alpha_diversity_stats:
    input:
        metadata_FP = "data/misc/processed_metadata.tsv",
        faith_pd_fp = "data/qiime/core_outputs/faith_pd.tsv",
        shannon_fp = "data/qiime/core_outputs/shannon_entropy.tsv"
    output:
        "stats/faith_total_results.tsv",
        "stats/faith_diet_results.tsv",
        "stats/shannon_total_results.tsv",
        "stats/shannon_diet_results.tsv"
    shell:
        """
        Rscript scripts/alpha_div_stats.R
        """


rule beta_diversity_plots:
    input:
        metadata_FP = "data/misc/processed_metadata.tsv",
        unweighted_FP = "data/qiime/core_outputs/unweighted_unifrac_pcoa_results.qza",
        weighted_FP = "data/qiime/core_outputs/weighted_unifrac_pcoa_results.qza"
    output:
        "plots/unweighted_unifrac_pcoa.pdf",
        "plots/weighted_unifrac_pcoa.pdf"
    shell:
        """
        Rscript scripts/beta_div_plots.R
        """


rule beta_diversity_stats:
    input:
        metadata_FP = "data/misc/processed_metadata.tsv",
        uw_dist_fp = "data/qiime/core_outputs/uw_dist_matrix.tsv",
        w_dist_fp = "data/qiime/core_outputs/w_dist_matrix.tsv"
    output:
        "stats/w_adonis_results.tsv",
        "stats/uw_adonis_results.tsv"
    shell:
        """
        Rscript scripts/beta_div_stats.R
        """


rule homogeneity:
    input:
        metadata_FP = "data/misc/processed_metadata.tsv",
        uu_dist_fp = "data/qiime/core_outputs/uw_dist_matrix.tsv",
        wu_dist_fp = "data/qiime/core_outputs/w_dist_matrix.tsv"
    output:
        "stats/wu_homogeneity.tsv",
        "stats/uu_homogeneity.tsv",
        "plots/wu_homogeneity.pdf",
        "plots/uu_homogeneity.pdf"
    shell:
        """
        Rscript scripts/homog_calc.R
        """


rule resiliency:
    input:
       metadata_FP = "data/misc/processed_metadata.tsv",
       uu_dist_fp = "data/qiime/core_outputs/uw_dist_matrix.tsv",
       wu_dist_fp = "data/qiime/core_outputs/w_dist_matrix.tsv"
    output:
        "stats/uu_resiliency.tsv",
        "stats/wu_resiliency.tsv",
        "plots/wu_resiliency.pdf",
        "plots/uu_resiliency.pdf"
    shell:
        """
        Rscript scripts/resil_calc.R
        """


rule family_abundance_plots:
    input:
        otu_table_FP = "data/qiime/taxonomy_filtered.qza",
        tax_FP = "data/qiime/taxonomy.qza",
        metadata_FP = "data/misc/processed_metadata.tsv"
    output:
        "plots/family_abun1.pdf",
        "plots/family_abun2.pdf"
    shell:
        """
        Rscript scripts/family_abun_plots.R
        """


rule family_abundance_stats:
    input:
       otu_table_FP = "data/qiime/taxonomy_filtered.qza",
       tax_FP = "data/qiime/taxonomy.qza",
       metadata_FP = "data/misc/processed_metadata.tsv"
    output:
        "stats/family_abun_lm.tsv"
    shell:
        """
        Rscript scripts/family_abun_stats.R
        """


rule histopathology:
    input:
        metadata_FP = "data/misc/processed_metadata.tsv",
        histo_FP = "data/misc/histo_data.csv"
    output:
        "plots/histopathology.pdf",
        "stats/histopathology_lm.tsv",
        "stats/histopathology_dunn.tsv"
    shell:
        """
        Rscript scripts/histopathology.R
        """


rule butyrate_plots:
    input:
        metadata_FP = "data/misc/processed_metadata.tsv",
        tax_FP = "data/qiime/taxonomy.qza",
        ko_contrib_FP = "data/picrust/tss3_meta_contrib.tsv"
    output:
        "plots/butyrate_kinase.pdf",
        "plots/butyryl_coa_transferase.pdf"
    shell:
        """
        Rscript scripts/butyrate_plots.R
        """


rule bile_acid_plots:
    input:
       metadata_FP = "data/misc/processed_metadata.tsv",
       tax_FP = "data/qiime/taxonomy.qza",
       ko_contrib_FP = "data/picrust/tss3_meta_contrib.tsv"
    output:
        "plots/baiH.pdf",
        "plots/baiI.pdf"
    shell:
        """
        Rscript scripts/bile_acid_plots.R
        """


rule butyrate_bile_stats:
    input:
       metadata_FP = "data/misc/processed_metadata.tsv",
       ko_contrib_FP = "data/picrust/tss3_meta_contrib.tsv"
    output:
        "stats/buty_enzyme_lm.tsv",
        "stats/bile_enzyme_lm.tsv"
    shell:
        """
        Rscript scripts/butyrate_bile_stats.R
        """