---
title: "tss_core_outputs"
author: "Madi"
date: "2023-05-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(qiime2R)
library(tidyverse)
library(cowplot)
library(magrittr)
library(vegan)
library(viridis)
library(microshades)
library(phyloseq)
library(ggh4x)
library(broom)
```

**Basic Functions**
```{r}
metadata_fixer <- function(metadata_fp) {
  tmpMeta <- read_tsv(metadata_fp, n_max = 2)
  mycols <- colnames(tmpMeta)
  metadata <- read_tsv(metadata_fp, skip = 2, col_names = mycols)
  names(metadata)[names(metadata) == '#SampleID'] <- 'sampleid'
  metadata %>% 
    filter(!is.na(diet)) %>% 
    mutate(day_post_inf = if_else(day_post_inf == 2, 3, day_post_inf)) %>% 
    mutate(diet = as.factor(diet)) -> metadata
  return(metadata)
}

## for faith's pd or any qiime output files that need the same second row exclusion as the metadata file
file_fixer <- function(metadata_fp) {
  tmpMeta <- read_tsv(metadata_fp, n_max = 2)
  mycols <- colnames(tmpMeta)
  metadata <- read_tsv(metadata_fp, skip = 2, col_names = mycols)
  names(metadata)[names(metadata) == '#SampleID'] <- 'sampleid'
  return(metadata)
}


## this function will pull out the percent variations from a specified column so you can add it to your pcoa plots 
pcoa_ax_lab <- function(unifrac_var, col_name){
  uni_lab <- as.character(round(unifrac_var[col_name] * 100, 2))
  uni_lab <- paste0(col_name, ' - ', uni_lab, '%')
  return(uni_lab)
}
```

**Microshades Functions**
```{r}
## phyloseq input file prep (bc phyloseq is really nit-picky)
## metadata prep 
## make sure that you change the 'read' functions depending on whether you
## have a 'tsv' or a 'csv' 
metadata_prep <- function(metadata_fp) {
  tmpMeta <- read_tsv(metadata_fp, n_max = 2)
  mycols <- colnames(tmpMeta)
  metadata <- read_tsv(metadata_fp, skip = 2, col_names = mycols)
  names(metadata)[names(metadata) == '#SampleID'] <- 'sampleid'
  metadata %>% 
    filter(!is.na(diet)) %>% 
    mutate(day_post_inf = if_else(day_post_inf == 2, 3, day_post_inf)) -> metadata
  metadata <- as.data.frame(metadata)
  rownames(metadata) <- metadata$sampleid
  return(metadata)
}

## otu table prep with missing samples 
## assumes that you're reading in a 'tsv' file 
otu_phyloseq_prep <- function(otu_fp, missing_samples){
  otu_table <- read_tsv(file = otu_fp)
  otu_table %>% 
    gather(-asv, key = sample, value = count) %>% 
    filter(sample %in% missing_sample) %>% 
    spread(sample, count) -> otu_missing
  names(otu_missing)[names(otu_missing) == 'asv'] <- ' '
  otu_missing <- as.matrix(otu_missing)
  class(otu_missing) <- "numeric"
  return(otu_missing)
}

## tax table prep with missing samples 
## assumes that its a 'qza' file from qiime2
tax_table_prep <- function(taxonomy_fp, wanted_asvs){
  tax_table <- read_qza(taxonomy_fp)$data 
  tax_table %>% 
    parse_taxonomy() %>% 
    as_tibble(rownames = 'asv') %>%
    filter(asv %in% wanted_asvs) -> tax_missing
  names(tax_missing)[names(tax_missing) == 'asv'] <- ' '
  return(tax_missing)
}

## this assumes that you're reading in a qiime2 'qza' file
## this also doesn't filter out anything for you 
## but you usually shouldn't need the filtration steps
otu_prep <- function(otu_fp){
  otu_table <- read_qza(file = otu_fp)$data
  return(otu_table)
}

## this assumes that your input is a 'qza' file 
tax_prep <- function(tax_fp, otu_table){
  tax_table <- read_qza(file = tax_fp)$data %>% 
    parse_taxonomy()
  tax_table %>% 
    filter(rownames(tax_table) %in% rownames(otu_table)) -> tax_table
  return(tax_table)
}

## generating color palettes based on the phyloseq object you generated
## generate pseq like normal since there's already a nice function for it 
## just returns the color_objs_GP for you, you'll need to extract the mdf/cdf for the next function 
mdf_cdf_gp <- function(pseq, tax_table, tax_rank, tax_subgroup){
  pseq %>% 
  tax_glom(tax_subgroup) %>% 
    phyloseq::transform_sample_counts(function(x) {
     x/sum(x)
      }) %>% 
    psmelt() %>% 
  filter(Abundance > 0) -> mdf_prep
  phylum_table <- tax_glom(pseq, taxrank=tax_rank, ) %>% otu_table()
  phyla.otunames <- rownames(phylum_table)
  phylums <- tax_table(as.matrix(tax_table))[phyla.otunames,tax_rank]
  sorted_phylums <- phylums[order(rowSums(phylum_table), decreasing=T)]

## can take ranked phylums out if you don't have issues with top abund. phylums being repeated
## ranked_phylums <- sort(unique(sorted_phylums[40:1]))
  color_objs_GP <- create_color_dfs(mdf_prep,
                                 subgroup_level = tax_subgroup,
                                 selected_groups = sorted_phylums[5:1])
  return(color_objs_GP)
}

##functions for taxa barplot construction via microshades 
## should be able to edit this like ggplot to add title and axes specs. 
tax_barplot <- function(mdf_GP, cdf_GP, diet_labs){
  barplot <- plot_microshades(mdf_GP, cdf_GP)
  barplot_1 <- barplot + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
    theme_bw(base_size = 18) +
    theme(legend.position = "none") +
    theme(axis.text.x = element_blank()) +
    facet_grid2(day_post_inf~diet, scales="free_x", independent = "x",
               labeller = labeller(diet = diet_labs)) +
    theme(legend.text = element_text(size = 8.5),
        strip.text.y = element_text(angle = 0))
  return(barplot_1)
}

tax_barplot_legend <- function(mdf_GP, cdf_GP, tax_subgroup){
  GP_legend <-custom_legend(mdf_GP, cdf_GP, 
                          legend_key_size=unit(0.4, "cm"),
                          legend_text_size=15, subgroup_level=tax_subgroup)
  return(GP_legend)
}

## putting entire taxa barplot together
whole_tax_barplot <- function(tax_barplot, tax_barplot_legend){
  tax_all <- plot_grid(tax_barplot, tax_barplot_legend,  rel_widths = c(1, .25),
                       align = 'tblr',
                       axis = 'tblr')
  return(tax_all)
}

##the big function!!
microshades_tax_barplot <- function(meta_fp, 
                                    otu_fp, 
                                    tax_fp,
                                    tax_rank, 
                                    tax_subgroup,
                                    title_text, 
                                    y_text, 
                                    facet_labels){
  pseq_meta <- metadata_prep(meta_fp)
  pseq_otu <- otu_prep(otu_fp)
  pseq_tax <- tax_prep(tax_fp, pseq_otu)
  pseq <- phyloseq(otu_table(pseq_otu, taxa_are_rows = TRUE),
                 tax_table(as.matrix(pseq_tax)),
                 sample_data(pseq_meta))
  mdf_and_cdf <- mdf_cdf_gp(pseq, pseq_tax, tax_rank, tax_subgroup)
  mdf_GP <- mdf_and_cdf$mdf
  cdf_GP <- mdf_and_cdf$cdf
  main_barplot <- tax_barplot(mdf_GP, cdf_GP, facet_labels) +
    ggtitle(title_text) +
    ylab(y_text)
  legend <- tax_barplot_legend(mdf_GP, cdf_GP, tax_subgroup)
  whole_barplot <- whole_tax_barplot(main_barplot, legend)
  return(whole_barplot)
}
```

**File paths for all needed files**
Doing this so that I don't have to worry about re-entering entire file paths all the time, they'll just be at the beginning
```{r}
metadata_FP <- '../data/misc/merged_metadata1.tsv'
seq_depth_FP <- '../data/misc/tss_seq_depth.tsv'
unweighted_fp <- '../data/qiime/core_outputs/unweighted_unifrac_pcoa_results.qza'
weighted_fp <- '../data/qiime/core_outputs/weighted_unifrac_pcoa_results.qza'
faith_pd_fp <- '../data/qiime/core_outputs/faith_pd.tsv'
shannon_fp <- '../data/qiime/core_outputs/shannon_entropy.tsv'
tax_fp <- '../data/qiime/taxonomy.qza'
sample_asv_fp <- '../data/misc/tss_sample_asv.tsv'


unwanted_samples <- c('Mock20220615A', 'Mock_1A', 'Mock_2A',
                      'Mock_3A', 'Mock_4A', 'Mock_5A', 'Mock_6A',
                      'Mock_7A', 'PCR Blank0',
                      'PCR Blank1', 'Mock_7', 'Mock_6',
                      'Mock_5', 'Mock_4', 'PCR blank')
```


**Reading in metadata file**
```{r}
metadata <- metadata_fixer(metadata_FP)
names(metadata)[names(metadata) == '#SampleID'] <- 'sampleid'

metadata %>% 
  select(sampleid, diet, study, day_post_inf) %>% 
  mutate(diet = as.factor(diet)) %>% 
  mutate(day_post_inf = if_else(day_post_inf == 2, 3, day_post_inf)) -> meta_red

```


**Reading in taxonomy file**
```{r}
## actual taxonomic file 
tss_tax <- read_qza(tax_fp)$data %>% 
                    parse_taxonomy() %>% 
                    rownames_to_column('asv')

## reading in file that has the asvs matched with sample ids so I can link the taxonomic info to 
## my core diversity outputs 
sample_asv <- read_tsv(sample_asv_fp)

## joining the two tables together 
tss_tax %>% 
  left_join(sample_asv, by = 'asv') -> big_tax
```


**Reading in weighted and unweighted unifrac files**
```{r}
## unweighted files 
tss_unweighted <- 
  read_qza(unweighted_fp)$data
##used for stats 
tss_unweighted_var <- tss_unweighted$ProportionExplained
##used for pcoa plot
tss_unweighted <- tss_unweighted$Vectors
names(tss_unweighted)[names(tss_unweighted) == 'SampleID'] <- 'sampleid'

## reading in faith's pd 
uni_faith_pd <- read_tsv(faith_pd_fp)
names(uni_faith_pd)[names(uni_faith_pd) == '#SampleID'] <- 'sampleid'

tss_unweighted %>% 
  left_join(meta_red, by = 'sampleid') %>% 
  left_join(uni_faith_pd, by = 'sampleid') -> tss_unweighted

## weighted files
tss_weighted <- 
  read_qza(weighted_fp)$data
##used for stats 
tss_weighted_var <- tss_weighted$ProportionExplained
##used for pcoa plot
tss_weighted <- tss_weighted$Vectors
names(tss_weighted)[names(tss_weighted) == 'SampleID'] <- 'sampleid'

tss_weighted %>% 
  left_join(meta_red, by = 'sampleid') %>% 
  left_join(uni_faith_pd, by = 'sampleid') -> tss_weighted
```

**Unweighted PcoA plot**
```{r, fig.height=3, fig.width=6, warning=FALSE}
uw_uni_xlab <- pcoa_ax_lab(tss_unweighted_var, 'PC1')
uw_uni_ylab <- pcoa_ax_lab(tss_unweighted_var, 'PC2')

diet_labs <- 
    c('Chow', 
      'High Fat / High Fiber', 
      'High Fat / Low Fiber', 
      'Low Fat / High Fiber', 
      'Low Fat / Low Fiber')

names(diet_labs) <- c('Chow', 'HF/HF', 'HF/LF', 'LF/HF', 'LF/LF')

tss_unweighted %>% 
  filter(!is.na(diet)) %>% 
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point(aes(fill = faith_pd), pch = 21, alpha = 0.7) +
  theme_bw(base_size = 14) +
  scale_fill_distiller(palette = 'Spectral', name = "Faith's PD") +
  facet_grid(day_post_inf~diet, 
             labeller = labeller(diet = diet_labs)) +
  theme(legend.text = element_text(size = 8.5),
        strip.text.y = element_text(angle = 0)) +
  ggtitle('Total Sum Scaled Unweighted UniFrac') +
  labs(x = uw_uni_xlab, y = uw_uni_ylab) -> tss_unweighted_unifrac

tss_unweighted_unifrac
  
```

**Weighted PcoA plot**
```{r, fig.height=3, fig.width=6, warning=FALSE}
w_uni_xlab <- pcoa_ax_lab(tss_weighted_var, 'PC1')
w_uni_ylab <- pcoa_ax_lab(tss_weighted_var, 'PC2')

diet_labs <- 
    c('Chow', 
      'High Fat / High Fiber', 
      'High Fat / Low Fiber', 
      'Low Fat / High Fiber', 
      'Low Fat / Low Fiber')

names(diet_labs) <- c('Chow', 'HF/HF', 'HF/LF', 'LF/HF', 'LF/LF')

tss_weighted %>%
  filter(!is.na(diet)) %>% 
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point(aes(fill = faith_pd), pch = 21, alpha = 0.7) +
  theme_bw(base_size = 14) +
  scale_fill_distiller(palette = 'Spectral', name = "Faith's PD") +
  facet_grid(day_post_inf~diet, 
             labeller = labeller(diet = diet_labs)) +
  theme(legend.text = element_text(size = 8.5),
        strip.text.y = element_text(angle = 0)) +
  ggtitle('Total Sum Scaled Weighted UniFrac') +
  labs(x = w_uni_xlab, y = w_uni_ylab) -> tss_weighted_unifrac

tss_weighted_unifrac
```


**Faith's Phylogenetic Diversity files**
so, there are about 10 samples that don't have metadata counterparts and both R and qiime2 are getting really mad about it and I don't know how to fix it. I can still run faith's PD in qiime2 to get an output but I haven't been able to do anything else. I'm slightly frustrated by this but it's fine (it's not fine)

**attempting to find the 10 missing samples in my metadata file**
I don't need this code anymore since I thought the missing samples meant something but they in fact do not (rip). 
```{r}
# faith_missing_fp <- '~/gut_microbiome_metabolomics/total_sum_scaled/tss_core_metrics2/faith_pd.tsv'
# tmp_faith <- read_tsv(faith_missing_fp, 
#                       n_max = 2)
# faith_cols <- colnames(tmp_faith)
# faith_missing <- read_tsv(faith_missing_fp, skip = 2, col_names = faith_cols)
# names(faith_missing)[names(faith_missing) == 'Sample ID'] <- 'sampleid'
# faith_sample <- faith_missing$sampleid
# 
# metadata %>% 
#   filter(!(sampleid %in% faith_sample)) -> missing_samples
# 
# missing_samples <- missing_samples$sampleid
# 
# metadata %>% 
#   filter(!(sampleid %in% missing_samples)) -> faith_meta

```

**Faith's PD plot (without taxonomic info)**
(finally!)
```{r, fig.height=5, fig.width=8, warning=FALSE}
## reading in faith's pd file
faith_pd_fp <- 
  '~/gut_microbiome_metabolomics/total_sum_scaled/tss3/tss_core_metrics3/faith_pd/data/alpha-diversity.tsv'
uni_faith_pd <- read_tsv(faith_pd_fp)
names(uni_faith_pd)[names(uni_faith_pd) == '#SampleID'] <- 'sampleid'

## joining it with the metadata file for plotting 
meta_red %>% 
  left_join(uni_faith_pd, by = 'sampleid') -> faith_pd

diet_labs <- 
    c('Chow', 
      'High Fat / High Fiber', 
      'High Fat / Low Fiber', 
      'Low Fat / High Fiber', 
      'Low Fat / Low Fiber')
names(diet_labs) <- c('Chow', 
                      'HF/HF', 
                      'HF/LF', 
                      'LF/HF', 
                      'LF/LF')

faith_pd %>%
  filter(!is.na(diet)) %>% 
  ggplot(aes(x = day_post_inf, y = faith_pd)) +
  geom_boxplot(aes(group = day_post_inf), outlier.shape = NA) +
  geom_jitter(width = 0.1, height = 0, alpha = 0.4) +
  scale_x_continuous(breaks = c(-15, -8, -3, 0, 3)) +
  geom_smooth(se = FALSE) +
  theme_bw(base_size = 16) +
  facet_grid(~diet, labeller = labeller(diet = diet_labs) ) +
  ggtitle("Total Sum Scaled Faith's Phylogenic Diversity") +
  xlab('Days Relative to Infection') +
  ylab("Faith's PD") -> total_faith_pd

total_faith_pd
```

**Faith's PD plot (with taxonomic info)**
I need to work on this nightmare (consult Elena's code for linking the taxonomic information to the alpha diversity)
```{r, warning=FALSE, fig.height=5, fig.width=10}
faith_pd %>% 
  select(sampleid, diet, study, day_post_inf, faith_pd) -> faith_red

## joining my faith pd values and metadata with taxonomic information for plot
faith_red %>% 
  left_join(big_tax, by = 'sampleid') -> faith_tax

## filtering out all low abundance classes of bacteria to focus on the big three
## trying to keep it consistent with by picrust output plots
wanted_class <- c('Bacilli', 'Bacteroidia', 'Clostridia')
faith_tax %>% 
  filter(Class %in% wanted_class) -> faith_class_filt

diet_labs <- 
    c('Chow', 
      'High Fat / High Fiber', 
      'High Fat / Low Fiber', 
      'Low Fat / High Fiber', 
      'Low Fat / Low Fiber')
names(diet_labs) <- c('Chow', 
                      'HF/HF', 
                      'HF/LF', 
                      'LF/HF', 
                      'LF/LF')

faith_class_filt %>%
  filter(!is.na(diet)) %>% 
  ##filter(!is.na(Class)) %>% 
  ggplot(aes(x = Class, y = faith_pd)) +
  ##geom_boxplot(aes(group = day_post_inf), outlier.shape = NA) +
  geom_jitter(width = 0.1, height = 0, alpha = 0.4) +
  scale_x_continuous(breaks = c(-15, -8, -3, 0, 3)) +
  theme_bw(base_size = 16) +
  facet_grid(~diet, labeller = labeller(diet = diet_labs) ) +
  ggtitle("Total Sum Scaled Faith's Phylogenic Diversity") +
  xlab('Class') +
  ylab("Faith's PD") -> tax_faith_pd

tax_faith_pd

```






**Microshades Taxa Barplot**
break up microshades function so that you get graphical outputs that you can edit rather than the whole barplot!
```{r, fig.width=25,fig.height=10}
## inputs
tss_metadata_fp <- 'merged_metadata1.tsv'
tss_otu_table_fp <- '~/gut_microbiome_metabolomics/total_sum_scaled/tss3/taxonomy_filtered.qza'
tss_tax_table_fp <- '~/gut_microbiome_metabolomics/total_sum_scaled/tss3/taxonomy.qza'
tss_tax_rank <- "Phylum"
tss_tax_subgroup <- "Family"
tss_title <- 'Total Sum Scaled Taxa Barplot'
tss_y <- 'Relative Abundance'
tss_diet_labs <- 
    c('Chow', 
      'High Fat / High Fiber', 
      'High Fat / Low Fiber', 
      'Low Fat / High Fiber', 
      'Low Fat / Low Fiber')
names(tss_diet_labs) <- c('Chow', 
                          'HF/HF', 
                          'HF/LF', 
                          'LF/HF', 
                          'LF/LF')

## barplot function
tss_barplot <- microshades_tax_barplot(tss_metadata_fp, 
                        tss_otu_table_fp, 
                        tss_tax_table_fp,
                        tss_tax_rank, 
                        tss_tax_subgroup,
                        tss_title, 
                        tss_y, 
                        tss_diet_labs) 

tss_barplot
```


**Example for how to read in my alpha diversity files (because it's been a struggle)**
**Shannon Entropy plot (without taxonomic info)**
```{r, fig.height=5, fig.width=8, warning=FALSE}
# another way to read in the data, I won't be using this one for my shannon plot though, it's just here for your reference 
# shannon <- read_qza('~/gut_microbiome_metabolomics/total_sum_scaled/tss_core_metrics2/shannon_vector2.qza')$data

## I'm going to use this one for graphical outputs 
shannon_fp <- '~/gut_microbiome_metabolomics/total_sum_scaled/tss3/tss_core_metrics3/shannon_entropy/data/alpha-diversity.tsv'
actual_shannon <- read_tsv(shannon_fp)

names(actual_shannon)[names(actual_shannon) == '...1'] <- 'sampleid'

## generating a biom table with the shannon entropy values and metadata
actual_shannon %>% 
  left_join(meta_red, by = 'sampleid') -> shannon_biom

## shannon entropy plot 
diet_labs <- 
    c('Chow', 
      'High Fat / High Fiber', 
      'High Fat / Low Fiber', 
      'Low Fat / High Fiber', 
      'Low Fat / Low Fiber')
names(diet_labs) <- c('Chow', 
                      'HF/HF', 
                      'HF/LF', 
                      'LF/HF', 
                      'LF/LF')

shannon_biom %>% 
  filter(!is.na(diet)) %>% 
  ggplot(aes(x = day_post_inf, y = shannon_entropy)) +
  geom_boxplot(aes(group = day_post_inf), outlier.shape = NA) +
  geom_jitter(width = 0.1, height = 0, alpha = 0.4) +
  scale_x_continuous(breaks = c(-15, -8, -3, 0, 3)) +
  geom_smooth(se = FALSE) +
  theme_bw(base_size = 16) +
  facet_grid(~diet, labeller = labeller(diet = diet_labs) ) +
  ggtitle("Total Sum Scaled Shannon Entropy") +
  xlab('Days Relative to Infection') +
  ylab("Shannon Entropy") -> shannon_entropy

shannon_entropy

```

**Shannon entropy plot (with taxonomic info)**
I also need to work on this plot, it's in the same place as the faith's pd taxonomic plot. 
```{r, warning=FALSE, fig.height=5, fig.width=8}
## joining shannon biom table with taxonomic information 
shannon_biom %>% 
  left_join(big_tax, by = 'sampleid') -> shannon_tax

## filtering out all low abundance classes of bacteria to focus on the big three
## trying to keep it consistent with by picrust output plots
wanted_class <- c('Bacilli', 'Bacteroidia', 'Clostridia')
shannon_tax %>% 
  filter(Class %in% wanted_class) -> shannon_class_filt

diet_labs <- 
    c('Chow', 
      'High Fat / High Fiber', 
      'High Fat / Low Fiber', 
      'Low Fat / High Fiber', 
      'Low Fat / Low Fiber')
names(diet_labs) <- c('Chow', 
                      'HF/HF', 
                      'HF/LF', 
                      'LF/HF', 
                      'LF/LF')

shannon_class_filt %>%
  filter(!is.na(diet)) %>% 
  ##filter(!is.na(Class)) %>% 
  ggplot(aes(x = day_post_inf, y = shannon_entropy)) +
  geom_boxplot(aes(group = day_post_inf), outlier.shape = NA) +
  geom_jitter(aes(fill = Class), width = 0.1, height = 0, alpha = 0.4, pch = 21) +
  geom_smooth(se = FALSE) +
  scale_fill_brewer(palette = 'Dark2') +
  scale_x_continuous(breaks = c(-15, -8, -3, 0, 3)) +
  theme_bw(base_size = 16) +
  facet_grid(~diet, labeller = labeller(diet = diet_labs) ) +
  ggtitle("Total Sum Scaled Shannon Entropy") +
  xlab('Days Relative to Infection') +
  ylab("Shannon Entropy") -> tax_shannon_plot

tax_shannon_plot
```


**Saving my outputs**
```{r}
## unweighted unifrac
ggsave("tss_unweighted_unifrac.pdf", 
       plot = tss_unweighted_unifrac,
       width = 11, 
       height = 5,
       path = '~/gut_microbiome_metabolomics/total_sum_scaled/tss3/plots')

## weighted unifrac
ggsave("tss_weighted_unifrac.pdf", 
       plot = tss_weighted_unifrac,
       width = 11, 
       height = 5,
       path = '~/gut_microbiome_metabolomics/total_sum_scaled/tss3/plots')

## faith's pd 
ggsave("tss_faith_pd.pdf", 
       plot = total_faith_pd,
       width = 11, 
       height = 6,
       path = '~/gut_microbiome_metabolomics/total_sum_scaled/tss3/plots')

## shannon
ggsave("tss_shannon.pdf", 
       plot = shannon_entropy,
       width = 11, 
       height = 6,
       path = '~/gut_microbiome_metabolomics/total_sum_scaled/tss3/plots')

## taxa barplot 
ggsave("tss_tax_barplot.pdf", 
       plot = tss_barplot,
       width = 25, 
       height = 15,
       path = '~/gut_microbiome_metabolomics/total_sum_scaled/tss3/plots')
```




